### ISSUE TITLE
[Discussion] Differentiate packages between packageLists

### ISSUE THREAD

**Issue by lmariscal** [nim-lang/nimble#517]
In my university projects my colleagues and I like using npm groups to be able to use private packages in nodejs, we found that using a private packageList has most of the same effect in Nim, the issue comes with packages with the same name, we could easily change the name in the packageList but personally as both of the packageLists grow this type of conflicts will occur more often.

I was thinking about making a PR to add a special character like in npm to differentiate between packageLists, npm uses '@' but in nimble it is being used for version selection, looking at the official list none of the packages has a '#' in the name which we could use to split between the packageList and the package name.

Do you have any thoughts on this implementation or on the especial character to split the packageList from the package name?

**Comment by dom96** [comment]
Interesting. So is this is what the ``@username/pkg`` that I've been seeing in Node means?

I would really love this in Nimble. But there is a lot of things to think about, one example off the top of my head:

* What happens when I install ``@dom96/strings`` and ``@cavariux/strings``?

**Comment by GULPF** [comment]
@dom96 Couldn't they be disambiguated with `import @dom96/strings` and `import @cavariux/strings`? If both are imported they would need to be assigned different aliases with `as`. In the file system it could be represented as folders named `@dom96` and `@cavariux` (this is how NPM does it).

One thing to consider: how would you keep track of who has authority over which prefix? NPM has a central register to keep track of this among other things, but Nimble doesn't.

**Comment by cavariux** [comment]
@dom96 to have no issues with prefix ownerships it would be convenient to keep the current flow of package creation for the Official packageList, which is the one that nimble would default for installation of packages if the user doesn't specify which package List to install from. As @GULPF commented each packageList would install on its own subdirectory except the Official one to avoid affecting current implementation and both import #cavariux/string and import #dom96/string would result without conflicts unless you use the same named method which nim would complain for ambiguity which can be resolved by appending the proper module.

The only issue that comes mind is how nim will react to this type of subdirectory imports.

**Comment by Araq** [comment]
So ... in order to avoid clashes every Nimble package has to have a name and in order to avoid Nimble package name clashes we introduce "prefix ownerships". And once these clash we introduce "super prefix ownerships"? Sounds silly to me, just keep Nimble package names unique.

**Comment by euantorano** [comment]
I may be misunderstanding here, but surely `@dom96` would represent the organisation/user that created the package, where as I would imagine that a `packageList` means something like `https://github.com/nim-lang/packages/blob/master/packages.json` (a list of packages).

I can personally see a couple of cases where you might have a private package list that has a packge in it that exists in the "official" list that you want to override the one from the "official" list (eg: you have an internal company verison of something like `jester` that adds some company-specific features).

**Comment by GULPF** [comment]
@euantorano I think the idea is that if you're in control of `@dom96` you can safely use private packages in the `@dom96` namespace without risking collisions.

If a company has a fork of `jester` that they want to publish as a private Nimble package they should give it a different name IMO. What if one of the projects that use the internal package list want to use the upstream version of `jester`? If they use the same name, it's impossible.

**Comment by euantorano** [comment]
@GULPF True. I think it's worth looking at what other package managers do here. I know that NuGet for .net allows you to use multiple package feeds (or package lists), but I'm not sure how it handles conflicts.
